### DISCUSSION TITLE
Hashicorp License Change

### DISCUSSION THREAD

**Discussion by meysam81** [opentofu/opentofu#431] [General]
The whole organization of OpenTF Foundation is here to address the license change in the Terraform (which is just one part of all the license changes in their suite of products).

Knowing that they also changed the ToS of their registry and OpenTF is planning to have its own registry soon, I wonder:

1. Until the OpenTF registry is up and running, can someone specify the GitHub repo URL of the provider in the form of `source: https://github.com/hashicorp/terraform-provider-aws` to avoid violating the ToS? (I tried with such a source, and it didn't work!)
2. If so, why would OpenTF go through the trouble of setting up a registry? If not, how will Hashicorp know about all the use cases of its binary if the organization's repository is private? My naive question is, will they track all the use cases? Or can someone dodge being detected while using it in violation of license and ToS?

**Comment by gtirloni** [comment]
Terraform currently only speaks to a [Terraform Registry](https://developer.hashicorp.com/terraform/registry/api-docs) endpoint to discover metadata about providers and cannot download from a raw HTTPS URL like that. When you specify the [source address](https://developer.hashicorp.com/terraform/language/providers/requirements#source-addresses), Terraform will figure out the registry's address, the namespace and the actual provider's name and retrieve the metadata from the API. It then downloads a bunch of files, which you can see in a `.terraform.lock.hcl` file.

I don't know if Hashicorp's TF Registry source code is public or bound to the new BSL license but it seems trivial to implement it and various companies associated with the OpenTF Foundation have indicated they're able to provide such service without much trouble.

If you want to avoid contacting Hashicorp's Registry, first disable Terraform's telemetry with `CHECKPOINT_DISABLE` (see [here](https://developer.hashicorp.com/terraform/cli/commands#disable_checkpoint)). Next you need to decide how to present the providers to Terraform.

You can use the provider_installation block to specify a local directory that follows some conventions specified [here](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation).

Alternatively, you can download the providers to the plugin cache directory and hope Terraform will use that and not contact Hashicorp. Two excellent answers you can read about this:
* https://stackoverflow.com/questions/66376560/why-is-terraform-pulling-a-provider-from-hashicorp-i-declare-nowhere
* https://stackoverflow.com/questions/50944395/use-pre-installed-terraform-plugins-instead-of-downloading-them-with-terraform-i

I hope once OpenTF creates their own registry, this will be less troublesome. Also expect OpenTF binary to be stripped of any phone home / telemetry code that would allow Hashicorp to track usage but it's understandably too early for that.
